### DISCUSSION TITLE
Deploy-Application.exe Pass a variable

### DISCUSSION THREAD

**Discussion by RalphConyers** [PSAppDeployToolkit/PSAppDeployToolkit#857] [General]
I have a script that needs to use either a Test or Production config file.  Instead of creating 2 separate install .ps1 files, I want to pass a variable when launching Deploy-Application.exe.  Is that possible?

**Comment by That-Annoying-Guy** [comment]
not sure. I try to avoid Deploy-Application.exe as much as possible. 
Deploy-Application.exe's purpose is to hide the PowerShell console from the user.

With Deploy-Application.exe out of the way you can call Deploy-Application.ps1 directly and use an additional _parameter_ to specify Test or Production. 

In your exact scenario, however, I avoid using a parameter and just detect which environment the script is running in and select the appropriate config file inside Deploy-Application.ps1

**Reply by RalphConyers** [reply]
huh... I tried to just use the .ps1 file but adding -DeploymentType and -DeployMode didn't work. Any other suggestions?

**Comment by jadelise** [comment]
Since you are calling the ps1, try using the PowerShell ARG variable to pass one or more variables of your own.  This allows you to add your variables to the command line.
This is not very clear because most of the documentation refers to this $ARGS as a function variable, but a script also has it when run as in the last example.

Good luck on your journey to PowerShell Mastery!


Use co-pilot and search for.

powershell args variable

You will receive the following  information
Searching for: powershell args variable
Generating answers for you...
The $args variable is an automatic variable in PowerShell that contains an array of values for undeclared parameters passed to a script or function <https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4> 1<https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4>2<https://www.delftstack.com/howto/powershell/accessing-the-args-array-in-powershell/>. For example, if you call a function with the following command: .\myFunction.ps1 -param1 value1 -param2 value2, the $args variable will contain an array of values value1 and value2 <https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4> 1<https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4>.
If you want to pass the $args variable to a function, you can simply include it as a parameter in the function definition. For example, the following code defines a function called main that takes $args as a parameter:
function main($args) {
    Write-Host $args
}
You can then call the main function and pass the $args variable as a parameter:
.\myScript.ps1 arg1 arg2 arg3
This will output the values of $args to the console.
I hope this helps! Let me know if you have any other questions.
Learn more
***@***.******@***.******@***.******@***.***<https://www.delftstack.com/zh/howto/powershell/accessing-the-args-array-in-powershell/>delftstack.com
1of30

**Reply by RalphConyers** [reply]
awesome 👍 this is very helpful... so dump PowerShell and this toolkit and back to batch scripts. :-)

**Comment by jadelise** [comment]
I am not sure I made my point.  The $ARG variable IS part of PowerShell, and the Toolkit script can use it just like any other PowerShell script.
So I don’t think I implied that we should dump PowerShell or the toolkit. On the contrary, a toolkit is a great tool that allows those with modest PowerShell skills to be very productive and increase their PowerShell skills.  Rock On  Toolkit

**Reply by RalphConyers** [reply]
well, the toolkit is good out of the box but my previous employer heavily modified it, and it was phenomenal. Passing custom arguments with deploy-Application.exe was normal there. I didn't realize it was something that was done by them and not a standard option with the toolkit. With that said, with my current employer and the turnaround time I have, it's better for me to just dump this and go back to batch scripts. Now, when time permits, I'll circle back around and try to modify the toolkit for my needs.

**Comment by mjr4077au** [comment]
We're tracking this issue here: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/issues/200
